src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -290,104 +290,50 @@ private class Emit(
           (llen > const(Int.MaxValue.toLong)).mux(
             Code._fatal("Array range cannot have more than MAXINT elements."),
             len := (llen < 0L).mux(0L, llen).toI),
+          srvb.start(len, init = true),
           i := 0,
-          srvb.start(len, init=true),
           Code.whileLoop(i < len,
             srvb.addInt(start),
             srvb.advance(),
             i := i + 1,
             start := start + step),
           srvb.offset))
-      case x@ArrayMap(a, name, body, elementTyp) =>
-        val tin = coerce[TArray](a.typ)
-        val tout = x.typ
-        val srvb = new StagedRegionValueBuilder(fb, tout)
-        val addElement = srvb.addIRIntermediate(tout.elementType)
-        val etiin = coerce[Any](typeToTypeInfo(tin.elementType))
-        val xa = fb.newLocal[Long]("am_a")
-        val xmv = mb.newBit()
-        val xvv = fb.newLocal(name)(etiin)
-        val i = fb.newLocal[Int]("am_i")
-        val len = fb.newLocal[Int]("am_len")
-        val out = fb.newLocal[Long]("am_out")
-        val bodyenv = env.bind(name -> (etiin, xmv, xvv))
-        val lmissing = new LabelNode()
-        val lnonmissing = new LabelNode()
-        val ltop = new LabelNode()
-        val lnext = new LabelNode()
-        val lend = new LabelNode()
-        val (doa, ma, va) = emit(a)
-        val (dobody, mbody, vbody) = emit(body, env = bodyenv)
 
-        (doa, ma, Code(
-          xa := coerce[Long](va),
-          len := TContainer.loadLength(region, xa),
-          i := 0,
-          srvb.start(len, init = true),
-          Code.whileLoop(i < len,
-            xmv := !tin.isElementDefined(region, xa, i),
-            xvv := xmv.mux(
-              defaultValue(tin.elementType),
-              region.loadIRIntermediate(tin.elementType)(tin.loadElement(region, xa, i))),
-            dobody,
-            mbody.mux(srvb.setMissing(), addElement(vbody)),
-            srvb.advance(),
-            i := i + 1),
-          srvb.offset))
-      case x@ArrayFilter(a, name, condition) =>
-        val t = x.typ
-        val srvb = new StagedRegionValueBuilder(fb, t)
-        val elementTypeInfoA = coerce[Any](typeToTypeInfo(t.elementType))
-
-        val xa = fb.newLocal[Long]("af_a")
-        val xmv = mb.newBit()
-        val xvv = fb.newLocal(name)(elementTypeInfoA)
-        val i = fb.newLocal[Int]("af_i")
+      case _: ArrayMap | _: ArrayFilter =>
```

Comment:
This code looks great.  I'll let @tpoterba do the detailed review, but I have one high level organizational comment:

It would be nice if the Array operators which deforest could all be treated symmetrically by Emit.  I think there are two cases:

 - there is a filter or flatMap, namely, you don't know the size of the array.  Then you have to build into a temporary ArrayBuilder.
 - there are only map and range and the length is the length of the first non-deforest-able expression

I think if emitArrayIterator returns an Optional[Code[Int]] which contains the length (if we have it) of the output, then you can build into a array builder or directly construct the array if possible.

Reply:
ok, done.

Reply:
Beautiful!  I'm happy, I'll leave the details to @tpoterba.